pkg/controller/topo/entity/reconciler.go

```diff
@@ -139,36 +182,79 @@ func (r *Reconciler) reconcileDelete(ctx context.Context, entity *v1beta1.Entity
+		case v1beta1.StateAdding, v1beta1.StateAdded:
+			entity.Status = &v1beta1.EntityStatus{State: v1beta1.StateRemoving}
+			err := r.client.Status().Update(context.TODO(), entity)
+			if err != nil {
+				log.Warnf("Failed to reconcile updating state of entity %s, %s, %s", entity.Name, entity.Namespace, err)
+				return reconcile.Result{}, err
+			}
+			return reconcile.Result{}, nil
+		case v1beta1.StateRemoving:
+			// Check if topo service is available
+			objectKey := types.NamespacedName{Namespace: entity.Namespace, Name: topoServiceName}
+			topoServiceObject := &corev1.Service{}
+			if err := r.client.Get(context.TODO(), objectKey, topoServiceObject); err != nil && k8serrors.IsNotFound(err) {
+				// Set the state to StateRemoved if topo service is not found.
+				entity.Status = &v1beta1.EntityStatus{State: v1beta1.StateRemoved}
+				if err := r.client.Status().Update(context.TODO(), entity); err != nil {
+					log.Warnf("Failed to reconcile updating state of entity %s, %s, %s", entity.Name, entity.Namespace, err)
+					return reconcile.Result{}, err
+				}
+				log.Warnf("Failed to find topo service %s in namespace %s, %s", topoServiceName, entity.Namespace, err)
+				return reconcile.Result{}, err
+			}
+			// Connect to the topology service
+			conn, err := grpc.ConnectService(r.client, entity.Namespace, topoServiceName)
+			if err != nil {
+				log.Warnf("Failed to reconcile deleting entity %s, %s, %s", entity.Name, entity.Namespace, err)
+				return reconcile.Result{}, err
+			}
+			defer conn.Close()
+			client := topo.NewTopoClient(conn)
+			// Delete the entity from the topology
+			if err := r.deleteEntity(ctx, entity, client); err != nil && !errors.IsNotFound(err) {
+				log.Warnf("Failed to reconcile deleting entity %s, %s, %s", entity.Name, entity.Namespace, err)
+				return reconcile.Result{}, err
+			}
+			entity.Status = &v1beta1.EntityStatus{State: v1beta1.StateRemoved}
+			err = r.client.Status().Update(context.TODO(), entity)
+			if err != nil {
+				log.Warnf("Failed to reconcile updating state of entity %s, %s, %s", entity.Name, entity.Namespace, err)
+				return reconcile.Result{}, err
+			}
+			return reconcile.Result{}, nil
+		case v1beta1.StateRemoved, v1beta1.StatePending:
+			log.Debugf("Entity %s is already removed or never been added to the topo store.", entity.Name)
+			k8s.RemoveFinalizer(entity, topoFinalizer)
+			if err := r.client.Update(context.TODO(), entity); err != nil {
+				log.Warnf("Failed to reconcile deleting entity %s, %s, %s", entity.Name, entity.Namespace, err)
+				return reconcile.Result{}, err
+			}
+			return reconcile.Result{}, nil
 		}
-		defer conn.Close()
-
-		client := topo.NewTopoClient(conn)
+	}
 
-		// Delete the entity from the topology
-		if err := r.deleteEntity(ctx, entity, client); err != nil {
+	// Once the entity has been deleted, remove the topology finalizer
+	if entity.Status.State == v1beta1.StatePending || entity.Status.State == v1beta1.StateRemoved {
```

Comment:
This block seems redundant with the block handling `Removed` inside the switch above.